File: 2024/03_Mull_It_Over/Day03.js
// Day 3: Mull It Over
// https://adventofcode.com/2024/day/3

const fs = require('fs');
var readInput = function (filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8').split('\n').filter(Boolean);
};

function part1(input) {
  const data = input.join('');
  const regex = /mul\((\d+),(\d+)\)/g;
  let match;
  let totalSum = 0;
  while ((match = regex.exec(data)) !== null) {
    const num1 = Number(match[1]);
    const num2 = Number(match[2]);
    const product = num1 * num2;
    totalSum += product;
  }
  return totalSum;
}

function part2(input) {
  const data = input.join('');
  const regex = /(do\(\)|don't\(\)|mul\((\d+),(\d+)\))/g;
  let match;
  let totalSum = 0;
  let doMul = 1;
  while ((match = regex.exec(data)) !== null) {
    if (match[0].startsWith('do')) {
      doMul = match[0] === "don't()" ? 0 : 1;
      continue;
    }
    const num1 = Number(match[2]);
    const num2 = Number(match[3]);
    const product = doMul * num1 * num2;
    totalSum += product;
  }
  return totalSum;
}

console.log(part1(readInput('Day03'))); // 161
console.log(part1(readInput('Day03_test'))); // 155955228

console.log(part2(readInput('Day03_part2'))); // 48
console.log(part2(readInput('Day03_test'))); // 100189366

File: 2024/03_Mull_It_Over/Day03.kt
// Day 3: Mull It Over
// https://adventofcode.com/2024/day/3

import kotlin.io.path.Path
import kotlin.io.path.readText
fun readInput(name: String) = Path("$name.txt").readText().trim().lines()

fun main() {
    fun prepareInput(input: List<String>): String {
        return input.joinToString("")
    }
    
    fun part1(input: List<String>): Int {
        val data = prepareInput(input)
        val regex = Regex("mul\\((\\d+),(\\d+)\\)")
        var totalSum = 0
        regex.findAll(data).forEach { matchResult ->
            val num1 = matchResult.groupValues[1].toInt()
            val num2 = matchResult.groupValues[2].toInt()
            val product = num1 * num2
            totalSum += product
        }
        return totalSum
    }

    fun part2(input: List<String>): Int {
        val data = prepareInput(input)
        val regex = Regex("(do\\(\\)|don't\\(\\)|mul\\((\\d+),(\\d+)\\))")
        var totalSum = 0
        var multiplyFactor = 1
        regex.findAll(data).forEach { matchResult ->
            val match = matchResult.groupValues[0]
            if (match.startsWith("do")) {
                multiplyFactor = if (match == "don't()") 0 else 1
                return@forEach
            }
            val num1 = matchResult.groupValues[2].toInt()
            val num2 = matchResult.groupValues[3].toInt()
            val product = multiplyFactor * num1 * num2
            totalSum += product
        }
        return totalSum
    }

    check(part1(readInput("Day03")) == 161)
    check(part2(readInput("Day03_part2")) == 48)

    val input = readInput("Day03_test")
    println(part1(input)) // 155955228
    println(part2(input)) // 100189366
}
